### PR TITLE
Image Url Optimisation using S3

### LINKED ISSUE #996: Get a better optimized content cache management
With the current MWoffliner and without `--useDownloadCache`, more than 2/3 of the whole process duration, MWoffliner optimised pictures, reducing the size of them to secure as smaller ZIM files as possible. This is too long, considering the amount hardware resources we have and the one release/month throughput we want to achieve.

If we use a download cache, things could be better because the cache keeps the optimised version of the pictures. But this has a few weaknesses:
* This can not be shared through different servers (because this is a filesystem based cache)
* There is a mix betweetn optimisation cache and download cache (this is semantically wrong and might bring further problems)
* The cache works also for text/pictures... and here we are mostly talking from pictures (caching the text might slow down the proces - to be confirmed).

Another point, it that this cache is obviously MWoffliner specific and we might need that kind of solution for all our other scrapers. So we probably need a better solution generic and network based.

### PR DESCRIPTION
**Pull request by bakshiutkarsha**
Fixes #996

Wasabi Plugin added
Uploading new images to wasabi
Checking if the image exists using etag, if yes, downloading it from wasabi and writing it to zim file.
If image doesn't exists, again following step 2.
Unit test cases for changes

### CODE REVIEW

File: src/util/s3.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            if ( err ) { reject(err);
+            } else { resolve(true); }
+        });
+    });
+}
+
+function bufferToStream(binary: Buffer) {
+    const readableInstanceStream = new Readable({
+      read() {
+        this.push(binary);
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/util/s3.ts]: There are some issues coming when I am doing one push.

File: src/Downloader.ts [hunk 2]
```diff
@@ -638,26 +653,42 @@ class Downloader {
...
-      } else if (shouldCompress) {
-        // logger.warn(`Failed to reduce file size after optimisation attempt [${requestOptions.url}]... Went from [${resp.data.length}] to [${compressed.length}]`);
+  private getContentCb = async(requestOptions: any, handler: any) => {
+    logger.log(`Downloading [${requestOptions.url}]`);
+    try {
+      if (await this.isImageUrl(requestOptions.url) && this.optimisationCacheUrl) {
+        this.urlStatusInS3(requestOptions, handler);
+      } else {
+        const resp = await axios(requestOptions);
+        const responseHeaders = resp.headers;
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: Typescript is not accepting direct resp.headers when given in result object. Also, the caller function expects it without a key name.

File: src/Downloader.ts [hunk 3]
```diff
@@ -638,26 +652,36 @@ class Downloader {
       });
   }
 
-  private async getContentCb(requestOptions: any, handler: any) {
-    logger.info(`Downloading [${requestOptions.url}]`);
-    try {
-      const resp = await axios(requestOptions);
-      const responseHeaders = resp.headers;
+  private async getBufferedData(resp: any): Promise<any> {
+    return resp.headers['content-type'].includes('image/') ? await imagemin.buffer(resp.data, imageminOptions) : resp.data;
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: Await was used prior to my code changes, and it is needed because without it, its returning Promise<pending>.

File: src/Downloader.ts [hunk 4]
```diff
@@ -675,6 +699,26 @@ class Downloader {
     }
   }
 
+  private async compressImageAndUploadToS3<T>(requestOptions: any, handler: any) {
+    const resp = await axios(requestOptions);
+    const responseHeaders = resp.headers;
+    const content = await this.getBufferedData(resp);
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: Here also await is needed.

File: src/Downloader.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: I did this change as well.

File: src/Downloader.ts [hunk 5]
```diff
@@ -151,6 +156,14 @@ class Downloader {
     }
   }
 
+  public isImageUrl = (url: string): boolean => {
+    return IMAGE_REGEX.exec(path.extname(url).toLowerCase()) ? true : false;
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: I am not sure whether i understood the comment correctly. Why exactly we want to change toLowerCase() here? I used it because it will convert all the uppercases to lowercase and then in regex we do not have to put different checks for uppercases values like JPG, GIF etc.

[the remark this answers is not in the document]
Comment (kelson42) [reply src/Downloader.ts]: This complexify the code without good reason and this is less efficient than the proper regex. See https://www.w3schools.com/jsref/jsref_regexp_i.asp

File: src/Downloader.ts [hunk 6]
```diff
@@ -151,6 +156,14 @@ class Downloader {
     }
   }
 
+  public isImageUrl (url: string): boolean {
+    return IMAGE_REGEX.exec(path.extname(url).toLowerCase()) ? true : false;
+  }
+
+  public isMimeTypeImage (mimetype: string): boolean {
+    return mimetype.includes('image/') ? true : false;
```
Comment (kelson42) [inline src/Downloader.ts]: Here as well a regex would be more efficient/quicker to execute.

File: src/util/s3.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        Body: bufferToStream(imageResp.data),
+    };
+    if (!imageResp['content-length']) {
+        options = {
+            partSize: 10 * 1024 * 1024,
+            queueSize: 1,
+        };
+    }
+    try {
+        s3Config.upload( params, options, function (err: any, data: any) {
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/util/s3.ts]: Yes, they have retry mechanism and the max retry count depends on service from which it is requested.

File: src/mwoffliner.lib.ts [hunk 8]
```diff
@@ -154,6 +158,16 @@ async function execute(argv: any) {
   populateFilesToDownload(redis.client);
   populateFilesToRetry(redis.client);
 
+  // Check for s3 creds
```
Comment (kelson42) [inline src/mwoffliner.lib.ts]: `S3` uppercased

File: test/unit/downloader.test.ts [hunk 9]
```diff
@@ -96,6 +100,64 @@ test('Downloader class', async (t) => {
...
+    const isSvgFile = await downloader.isImageUrl('https://bm.wikipedia.org/static/images/project-logos/bmwiki-2x.svg');
+    t.assert(isSvgFile, 'Checked Image type: svg');
+
+    const isJpegFile = await downloader.isImageUrl('https://bm.wikipedia.org/static/images/project-logos/bmwiki-2x.JPEG');
+    t.assert(isJpegFile, 'Checked Image type: jpeg');
+
+    const isgifFile = await downloader.isImageUrl('https://bm.wikipedia.org/static/images/project-logos/bmwiki-2x.gif');
+    t.assert(isgifFile, 'Checked Image type: gif');
+
+    const isnotImage = await downloader.isImageUrl('https://en.wikipedia.org/w/api.php?action=query&meta=siteinfo&format=json');
```
Comment (kelson42) [inline test/unit/downloader.test.ts]: This are only the simple cases, unit test should test edge case with:
* undef value
* empty string value
* no extension
* maybe extension pattern at the end like "asdfasdf.webm.png" and "asdfasdf.png.svg".
* uppercased extension
*

File: src/Downloader.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/Downloader.ts]: ok, got it.

File: src/util/s3.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kelson42) [reply src/util/s3.ts]: So here how often would that be retried? Where do we change/configure that?

File: src/util/s3.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        };
+
+        return new Promise((resolve, reject) => {
+            this.s3Config.getObject(params, async (err: any, val: any) => {
+                if (err && err.statusCode === 404) {
+                    resolve();
+                } else {
+                    const valHeaders = (({ Body, ...o }) => o)(val);
+                    const urlHeaders = await axios.head(filepath);
+                    // Check if e-tag is in sync
```
Comment (kelson42) [inline src/util/s3.ts]: `ETag` not `e-tag`, You can keep 'etag' as property... but better proper naming in the comment.

File: src/util/s3.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/util/s3.ts]: default value of retry is 3.
however we can change while instantiate the S3 obj like this:
var s3 = new AWS.S3({maxRetries: 10});

File: src/s3.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        const params = {
+            Bucket: this.bucketName,
+            Key: path.basename(key),
+            Metadata: {etag: eTag },
+            Body: this.bufferToStream(data),
+        };
+
+        try {
+            this.s3Config.upload( params, function (err: any, data: any) {
+                if (data) {
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/s3.ts]: Now, I have kept the if block to check the error only.

File: test/unit/downloader.test.ts [hunk 12]
```diff
@@ -96,6 +100,73 @@ test('Downloader class', async (t) => {
...
+    const downloader = new Downloader({ mw, uaString: '', speed: 1, reqTimeout: 1000 * 60, useDownloadCache: true, downloadCacheDirectory: cacheDir, noLocalParserFallback: false, optimisationCacheUrl: 'random-string' , s3});
+
+    await s3.initialise();
+
+    const testImage = 'https://bm.wikipedia.org/static/images/project-logos/bmwiki-2x.png';
+    // Test for image where etag is not present
+    const etagNotPresent = await downloader.downloadContent(`https://en.wikipedia.org/w/extensions/WikimediaBadges/resources/images/badge-silver-star.png?70a8c`);
+    t.equals(etagNotPresent.responseHeaders.etag, undefined , 'Etag Not Present');
+
+    // FLOW OF IMAGE CACHING
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply test/unit/downloader.test.ts]: It is already another function called "Downloader class with optimisation". I am not sure what you meant exactly?
